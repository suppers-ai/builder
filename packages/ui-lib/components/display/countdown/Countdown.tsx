import { BaseComponentProps } from "../../types.ts";

export interface TimeLeft {
  days: number;
  hours: number;
  minutes: number;
  seconds: number;
  totalSeconds: number;
}

export interface CountdownProps extends BaseComponentProps {
  targetDate?: string | Date;
  showLabels?: boolean;
  showDays?: boolean;
  showHours?: boolean;
  showMinutes?: boolean;
  showSeconds?: boolean;
  size?: "xs" | "sm" | "md" | "lg" | "xl";
  // Controlled mode props
  timeLeft?: TimeLeft;
  onComplete?: () => void;
  onTick?: (timeLeft: TimeLeft) => void;
}

export function Countdown({
  class: className = "",
  targetDate,
  showLabels = true,
  showDays = true,
  showHours = true,
  showMinutes = true,
  showSeconds = true,
  size = "md",
  timeLeft,
  id,
  ...props
}: CountdownProps) {
  // Use controlled mode if timeLeft is provided, otherwise calculate from targetDate
  let currentTimeLeft: TimeLeft;

  if (timeLeft) {
    // Controlled mode - use provided timeLeft
    currentTimeLeft = timeLeft;
  } else if (targetDate) {
    // Uncontrolled mode - calculate from targetDate
    const target = typeof targetDate === "string" ? new Date(targetDate) : targetDate;
    const now = new Date();
    const timeDiff = Math.max(0, target.getTime() - now.getTime());

    const days = Math.floor(timeDiff / (1000 * 60 * 60 * 24));
    const hours = Math.floor((timeDiff % (1000 * 60 * 60 * 24)) / (1000 * 60 * 60));
    const minutes = Math.floor((timeDiff % (1000 * 60 * 60)) / (1000 * 60));
    const seconds = Math.floor((timeDiff % (1000 * 60)) / 1000);
    const totalSeconds = Math.floor(timeDiff / 1000);

    currentTimeLeft = { days, hours, minutes, seconds, totalSeconds };
  } else {
    // Fallback - all zeros
    currentTimeLeft = { days: 0, hours: 0, minutes: 0, seconds: 0, totalSeconds: 0 };
  }

  const textSizeClasses = {
    xs: "text-xs",
    sm: "text-sm",
    md: "text-base",
    lg: "text-lg",
    xl: "text-xl",
  };

  const fontSizeClasses = {
    xs: "font-mono text-xs",
    sm: "font-mono text-sm",
    md: "font-mono text-2xl",
    lg: "font-mono text-4xl",
    xl: "font-mono text-6xl",
  };

  return (
    <div class={`flex gap-2 ${className}`} id={id} {...props}>
      {showDays && (
        <div class="flex flex-col items-center">
          <span class={`${fontSizeClasses[size]} font-bold`}>
            {currentTimeLeft.days.toString().padStart(2, "0")}
          </span>
          {showLabels && <span class={`${textSizeClasses[size]} text-base-content/70`}>days</span>}
        </div>
      )}

      {showDays && (showHours || showMinutes || showSeconds) && (
        <span class={`${fontSizeClasses[size]} font-bold text-base-content/50`}>:</span>
      )}

      {showHours && (
        <div class="flex flex-col items-center">
          <span class={`${fontSizeClasses[size]} font-bold`}>
            {currentTimeLeft.hours.toString().padStart(2, "0")}
          </span>
          {showLabels && <span class={`${textSizeClasses[size]} text-base-content/70`}>hours</span>}
        </div>
      )}

      {showHours && (showMinutes || showSeconds) && (
        <span class={`${fontSizeClasses[size]} font-bold text-base-content/50`}>:</span>
      )}

      {showMinutes && (
        <div class="flex flex-col items-center">
          <span class={`${fontSizeClasses[size]} font-bold`}>
            {currentTimeLeft.minutes.toString().padStart(2, "0")}
          </span>
          {showLabels && (
            <span class={`${textSizeClasses[size]} text-base-content/70`}>minutes</span>
          )}
        </div>
      )}

      {showMinutes && showSeconds && (
        <span class={`${fontSizeClasses[size]} font-bold text-base-content/50`}>:</span>
      )}

      {showSeconds && (
        <div class="flex flex-col items-center">
          <span class={`${fontSizeClasses[size]} font-bold`}>
            {currentTimeLeft.seconds.toString().padStart(2, "0")}
          </span>
          {showLabels && (
            <span class={`${textSizeClasses[size]} text-base-content/70`}>seconds</span>
          )}
        </div>
      )}
    </div>
  );
}
