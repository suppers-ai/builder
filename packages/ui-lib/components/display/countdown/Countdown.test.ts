import { assertEquals, assertStringIncludes } from "https://deno.land/std@0.224.0/assert/mod.ts";
import { assertSnapshot } from "https://deno.land/std@0.224.0/testing/snapshot.ts";
import { renderToString } from "preact-render-to-string";
import { Countdown } from "./Countdown.tsx";

Deno.test("Countdown - basic rendering", () => {
  const html = renderToString(Countdown({
    timeLeft: { days: 1, hours: 2, minutes: 3, seconds: 42, totalSeconds: 94742 },
  }));
  assertStringIncludes(html, "flex gap-2");
  assertStringIncludes(html, "font-mono");
  assertStringIncludes(html, "42");
});

Deno.test("Countdown - with days", () => {
  const html = renderToString(Countdown({
    timeLeft: { days: 5, hours: 12, minutes: 30, seconds: 45, totalSeconds: 518445 },
  }));
  assertStringIncludes(html, "flex gap-2");
  assertStringIncludes(html, "05");
  assertStringIncludes(html, "12");
  assertStringIncludes(html, "30");
  assertStringIncludes(html, "45");
});

Deno.test("Countdown - with labels", () => {
  const html = renderToString(Countdown({
    timeLeft: { days: 0, hours: 24, minutes: 0, seconds: 0, totalSeconds: 86400 },
    showLabels: true,
  }));
  assertStringIncludes(html, "flex gap-2");
  assertStringIncludes(html, "hours");
});

Deno.test("Countdown - large size", () => {
  const html = renderToString(Countdown({
    timeLeft: { days: 0, hours: 1, minutes: 0, seconds: 0, totalSeconds: 3600 },
    size: "lg",
  }));
  assertStringIncludes(html, "text-4xl");
});

Deno.test("Countdown - medium size", () => {
  const html = renderToString(Countdown({
    timeLeft: { days: 0, hours: 0, minutes: 30, seconds: 0, totalSeconds: 1800 },
    size: "md",
  }));
  assertStringIncludes(html, "text-2xl");
});

Deno.test("Countdown - small size", () => {
  const html = renderToString(Countdown({
    timeLeft: { days: 0, hours: 0, minutes: 0, seconds: 15, totalSeconds: 15 },
    size: "sm",
  }));
  assertStringIncludes(html, "text-sm");
});

Deno.test("Countdown - with custom class", () => {
  const html = renderToString(Countdown({
    timeLeft: { days: 0, hours: 0, minutes: 0, seconds: 99, totalSeconds: 99 },
    class: "custom-countdown",
  }));
  assertStringIncludes(html, "custom-countdown");
});

Deno.test("Countdown - bordered style", () => {
  const html = renderToString(Countdown({
    timeLeft: { days: 0, hours: 0, minutes: 1, seconds: 28, totalSeconds: 88 },
    class: "border",
  }));
  assertStringIncludes(html, "border");
});

Deno.test("Countdown - with background", () => {
  const html = renderToString(Countdown({
    timeLeft: { days: 0, hours: 0, minutes: 1, seconds: 17, totalSeconds: 77 },
    class: "bg-neutral",
  }));
  assertStringIncludes(html, "bg-neutral");
});

Deno.test("Countdown - HTML snapshot", async (t) => {
  const html = renderToString(Countdown({
    timeLeft: { days: 0, hours: 0, minutes: 42, seconds: 0, totalSeconds: 2520 },
    showLabels: true,
  }));
  await assertSnapshot(t, html);
});

Deno.test("Countdown - HTML snapshot full format", async (t) => {
  const html = renderToString(Countdown({
    timeLeft: { days: 7, hours: 23, minutes: 59, seconds: 30, totalSeconds: 690570 },
    showLabels: true,
    size: "lg",
    class: "border",
  }));
  await assertSnapshot(t, html);
});
