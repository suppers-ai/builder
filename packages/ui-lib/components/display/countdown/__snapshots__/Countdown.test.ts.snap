export const snapshot = {};

snapshot[`CountdownDisplay - HTML snapshot 1`] = `'<div class="flex gap-2 "><div class="flex flex-col items-center"><span class="font-mono text-2xl font-bold">00</span><span class="text-base text-base-content/70">days</span></div><span class="font-mono text-2xl font-bold text-base-content/50">:</span><div class="flex flex-col items-center"><span class="font-mono text-2xl font-bold">00</span><span class="text-base text-base-content/70">hours</span></div><span class="font-mono text-2xl font-bold text-base-content/50">:</span><div class="flex flex-col items-center"><span class="font-mono text-2xl font-bold">42</span><span class="text-base text-base-content/70">minutes</span></div><span class="font-mono text-2xl font-bold text-base-content/50">:</span><div class="flex flex-col items-center"><span class="font-mono text-2xl font-bold">00</span><span class="text-base text-base-content/70">seconds</span></div></div>'`;

snapshot[`CountdownDisplay - HTML snapshot full format 1`] = `'<div class="flex gap-2 border"><div class="flex flex-col items-center"><span class="font-mono text-4xl font-bold">07</span><span class="text-lg text-base-content/70">days</span></div><span class="font-mono text-4xl font-bold text-base-content/50">:</span><div class="flex flex-col items-center"><span class="font-mono text-4xl font-bold">23</span><span class="text-lg text-base-content/70">hours</span></div><span class="font-mono text-4xl font-bold text-base-content/50">:</span><div class="flex flex-col items-center"><span class="font-mono text-4xl font-bold">59</span><span class="text-lg text-base-content/70">minutes</span></div><span class="font-mono text-4xl font-bold text-base-content/50">:</span><div class="flex flex-col items-center"><span class="font-mono text-4xl font-bold">30</span><span class="text-lg text-base-content/70">seconds</span></div></div>'`;
